Editorial + java implementation for CSES Range XOR Queries (#2099)
cpinitiative/usaco-guide: solutions/cses-1650.mdx

id: cses-1650
source: CSES
title: Range Xor Queries
author: Danh Ta Chi Thanh, Nathan Gong
---

## Explanation

Whenever we come across range queries, prefix sums should immediately come to

> mind. However, since we are dealing with range XOR, we have to use a prefix XOR

Reviewer: ```suggestion
mind. However, since the problem queries for XOR sums over a range, we have to use a prefix XOR
```

array rather than a prefix sum array. To do so, we can define $\texttt{prefix}$
as an array such that
$\texttt{prefix}[i]= \texttt{arr}[0] \oplus \texttt{arr}[1] \oplus \dots \oplus \texttt{arr}[i]$.
Thus, for each query $(a, b)$, our answer is simply
$\texttt{prefix}[a-1]\oplus \texttt{prefix}[b]$.

### Proof

We know that
$\texttt{prefix}[a-1]= \texttt{arr}[0] \oplus \texttt{arr}[1] \oplus \dots \oplus \texttt{arr}[a-1]$.
We also know that
$\texttt{prefix}[b]= \texttt{arr}[0] \oplus \texttt{arr}[1] \oplus \dots \oplus \texttt{arr}[b]$.

> Therefore:

Reviewer: ```suggestion
Therefore, since $a \oplus a = 0$:
```

Author: I think using `a` as a number might be misleading because it was already used in the explanation. I'll create another commit for this

Reviewer 2: ```suggestion
Therefore, since $x\oplus x = 0$ for any $x$:
```

$\texttt{prefix}[a-1] \oplus \texttt{prefix}[b]= \underbrace{\texttt{arr}[0] \oplus \texttt{arr}[0] \oplus \texttt{arr}[1] \oplus \texttt{arr}[1] \oplus \dots \oplus \texttt{arr}[a-1] \oplus \texttt{arr}[a-1]}_{\text{cancels out to 0}} \oplus \texttt{arr}[a] \oplus \texttt{arr}[a+1] \dots \oplus \texttt{arr}[b]$.

This then simplifies to
$\texttt{arr}[a] \oplus \texttt{arr}[a+1] \dots \oplus \texttt{arr}[b]$, which
is exactly what we're looking for.

## Implementation

**Time Complexity:** $\mathcal{O}(n+q)$

<LanguageSection>

<CPPSection>

```cpp
#include <bits/stdc++.h>

> 

Reviewer: ```suggestion
```

using namespace std;

> const int maxn = 2e5 + 10;

Reviewer 3: rename to MAX_N and change value to just 2e5

Reviewer 3: if 1-indexing is a problem just declare arrays like `arr[MAX_N + 1]`

> 

Reviewer: ```suggestion
```

> long long n, q, arr[maxn], prefix[maxn], a, b;

Reviewer 2: [optional] change to `int`

void solve() {

> 	cin >> n >> q;

Reviewer 3: wait why not just put this in main there's literally no reason for it to be in a separate method

> 	// arr and prefix use 1-based indexing

Reviewer: ```suggestion
	// make arr and prefix to use 1-based indexing
```

Reviewer 3: this isn't really needed imo

	for (int i = 1; i <= n; ++i) {
		cin >> arr[i];
	}
	// Build prefix XOR array
	prefix[1] = arr[1];
	for (int i = 2; i <= n; i++) {

> 		prefix[i] = prefix[i - 1] ^ arr[i];

Reviewer: ```suggestion
		// ^ is the XOR operator in C++
		prefix[i] = prefix[i - 1] ^ arr[i]; 
```

Author: I just added it as a note in the editorial

	}

	while (q--) {
		cin >> a >> b;
		cout << (prefix[a - 1] ^ prefix[b]) << endl;
	}
	return;
}

int main() {
	ios_base::sync_with_stdio(false);

> 	cin.tie(nullptr);

Reviewer 4: can you change this to ```cin.tie(0)```. I know both works but ```cin.tie(0)``` seem to be more common and accepted as the standard.

	solve();

> 	return 0;

Reviewer 3: ```suggestion
```

}
```

</CPPSection>

<JavaSection>

```java
import java.util.*;
import java.io.*;

public class RangeXorQueries {
	public static void main(String[] args) {
		Kattio io = new Kattio();
		int n = io.nextInt();
		int q = io.nextInt();

		// arr and prefix use 1-based indexing
		int[] arr = new int[n + 1];
		for (int i = 1; i <= n; i++) {
			arr[i] = io.nextInt();

> 		}
> 		// Build prefix XOR array

Reviewer: ```suggestion
		}
		
		// Build prefix XOR array
```

Reviewer 3: the whitespace isn't needed

		int[] prefix = new int[n + 1];
		prefix[1] = arr[1];
		for (int i = 2; i <= n; i++) {

> 			prefix[i] = prefix[i - 1] ^ arr[i];

Reviewer: ```suggestion
			// ^ is the XOR operator in C++
			prefix[i] = prefix[i - 1] ^ arr[i];
```

Reviewer 3: uuuhhh wrong language?

		}

		for (int i = 0; i < q; i++) {
			int a = io.nextInt();
			int b = io.nextInt();
			io.println(prefix[a - 1] ^ prefix[b]);
		}
		io.close();
	}

	// CodeSnip{Kattio}
}
```

> 

Reviewer: ```suggestion
```

Author: This whitespace is to match the format given in the adding solutions module

id: cses-1650
source: CSES
title: Range Xor Queries
author: Danh Ta Chi Thanh, Nathan Gong
---

> [Unofficial Editorial (C++)](https://codeforces.com/blog/entry/77128?#comment-637536)

Reviewer 2: ```suggestion
```
I don't see a reason for including this.

## Explanation

Whenever we come across range queries, prefix sums should immediately come to
mind. However, since we are dealing with range XOR, we have to use a prefix XOR
array rather than a prefix sum array. To do so, we can define $\texttt{prefix}$
as an array such that
$\texttt{prefix}[i]= \texttt{arr}[0] \oplus \texttt{arr}[1] \oplus \dots \oplus \texttt{arr}[i]$.

> Thus, for each query, our answer is simply

Reviewer 2: ```suggestion
Thus, for each query $(a, b)$, our answer is simply
```
